course-content/22_cleaning_it_up.qmd

```diff
@@ -0,0 +1,410 @@
+
+```{python}
+techhealth = "https://raw.githubusercontent.com/bcgov/" \
+        "ds-intro-to-python/main/data/techhealth.csv"
+m_health = pd.read_csv(techhealth)
+m_health.head()
+```
+
+<br>
+This abridged version of the dataset contains 24 cases and 13 variables:
+
+- Timestamp
+- Age
+- Gender
+- Country
+- Self Employed: Are you self-employed?
+- Family History: Do you have a family history of mental illness?
+- Treatment: Have you sought treatment for a mental health condition?
+- Work Interfere: If you have a mental health condition, do you feel that it interferes with your work?
+- Remote Work: Do you work remotely (outside of an office) at least 50% of the time?
+- Tech Company: Is your employer primarily a tech company/organization?
+- Benefits: Does your employer provide mental health benefits?
+- Leave: How easy is it for you to take medical leave for a mental health condition?
+- Mental Health Consequence: Do you think that discussing a mental health issue with your employer would have negative consequences?
+
+
+### Cleaning the header row
+
+A good practice, both in terms of ensuring a consistent convention for column names and for the ability to write more efficient code, is to ensure that all column header names do not have spaces in them and are all in lower case. To do this with our current dataset, we use the `str.replace()` function on all the header row, identified in pandas via .columns:
+
+
+```{python}
+m_health.columns = m_health.columns.str.replace(' ', '_')
+m_health.columns = m_health.columns.str.lower()
+m_health.head(10)
+m_health.info()
+```
+
+The other thing to notice that there are no missing values! So we can just concentrate on other cleaning tasks. 
+
+
+
+### Ensuring the right datatypes
+
+A good practice is to notice whether there are series in the dataframe that are improperly datatyped (that is, where the "Dtype" indicated does not correspond to the datatype that it should be). 
+
+Taking a look at the data, we notice that "Timestamp" is datatyped as an "object", which in this context means a "string". However, that is sub-optimal because the content is time-oriented, and by coding it as an object/string, we limit a lot of its value. If we identify datatype as a time oriented one, that can give us advantages downstream in being able to work with the data more elegantly.  
+
+Pandas has functions specifically designed to cast datatypes from one to another, as was covered in the previous section with the `astype()` function.  
+
+To cast a variable as a datetime datatype, we can use the pandas `to_datetime()` function and pass in the pandas series you want to modify:
+
+```{python}
+m_health['timestamp'] = pd.to_datetime(m_health['timestamp'])
+m_health.head()
+m_health.info()
+```
+
+Et voila - timestamp is now of the datatime64 datatype! Were we wanting to leverage timestamp's date time qualities, we could now do so more easily. 
```

Comment:
Because dates are tricky/annoying - I would add a comment that this only works in such a default fashion with data that is nicely formatted. If the timestamp column has different date types, or isn't actually a date column at all, this method raises exceptions. Maybe just showing how it raises an error if you try it on age or gender or something fails.